In [ ]:
import pandas as pd
import numpy as np

In [ ]:
datos = pd.read_csv("Dataset.csv",sep=";")

In [ ]:
datos.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

# Analisis General

## Exploración Inicial

In [ ]:
print(datos.shape)

(20739, 17)


In [ ]:
print(datos.head())

   age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no  


In [ ]:
print(datos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20739 entries, 0 to 20738
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        20739 non-null  int64 
 1   job        20739 non-null  object
 2   marital    20739 non-null  object
 3   education  20739 non-null  object
 4   default    20739 non-null  object
 5   balance    20739 non-null  int64 
 6   housing    20739 non-null  object
 7   loan       20739 non-null  object
 8   contact    20739 non-null  object
 9   day        20739 non-null  int64 
 10  month      20739 non-null  object
 11  duration   20739 non-null  int64 
 12  campaign   20739 non-null  int64 
 13  pdays      20739 non-null  int64 
 14  previous   20739 non-null  int64 
 15  poutcome   20739 non-null  object
 16  y          20739 non-null  object
dtypes: int64(7), object(10)
memory usage: 2.7+ MB
None


Parece no haber nulos, ya que todas las variables tienen 20739 valores, ahora estadisticas descriptivas de las variables númericas

In [ ]:
print(datos.describe())

                age        balance           day      duration      campaign  \
count  20739.000000   20739.000000  20739.000000  20739.000000  20739.000000   
mean      41.095762    1148.639230     15.976469    259.027243      3.089541   
std       11.251527    4501.290387      8.588330    261.613962      3.660176   
min       20.000000   -8019.000000      1.000000  -1389.000000      1.000000   
25%       33.000000      30.000000      8.000000    103.000000      1.000000   
50%       40.000000     350.000000     15.000000    178.000000      2.000000   
75%       49.000000    1191.000000     23.000000    317.000000      3.000000   
max      530.000000  527532.000000     31.000000   3881.000000     63.000000   

              pdays      previous  
count  20739.000000  20739.000000  
mean      -0.990935      0.000241  
std        1.305461      0.034720  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       -1.000000      0.

Se puede ver algunas inconsistencias como el "age" maximo 530, numeros negativos en "duration". Se analizara detalladamente después.

In [ ]:
# Revisa valores nulos
print(datos.isnull().sum())

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64


## Duplicados

In [ ]:
# Filas duplicadas
datos[datos.duplicated(keep=False)]

age          job  marital  education default  balance housing loan  \
523    43  blue-collar  married  secondary     yes       -7      no   no   
1201   43  blue-collar  married  secondary     yes       -7      no   no   

      contact  day month  duration  campaign  pdays  previous poutcome   y  
523   unknown    8   may        70         1     -1         0  unknown  no  
1201  unknown    8   may        70         1     -1         0  unknown  no

In [ ]:
# Eliminando filas duplicadas 
datos.drop_duplicates(inplace=True)

# Analisis por variables

## Age

In [ ]:
datos["age"].describe()

count    20738.000000
mean        41.095670
std         11.251791
min         20.000000
25%         33.000000
50%         40.000000
75%         49.000000
max        530.000000
Name: age, dtype: float64

In [ ]:
datos["age"].describe()

count    20738.000000
mean        41.095670
std         11.251791
min         20.000000
25%         33.000000
50%         40.000000
75%         49.000000
max        530.000000
Name: age, dtype: float64

In [ ]:
datos[datos["age"]>70]

age            job   marital  education default  balance housing loan  \
229   530       services  divorced  secondary      no     1467     yes   no   
1076  490       services    single    primary      no      114     yes   no   
2229  123        student    single   tertiary      no      250     yes   no   
3018  332         admin.    single  secondary      no       91     yes   no   
5799  311  self-employed    single  secondary      no      360     yes   no   
6053  399    blue-collar   married    primary      no     2805     yes   no   

      contact  day month  duration  campaign  pdays  previous poutcome   y  
229   unknown    5   may       124         1     -1         0  unknown  no  
1076  unknown    7   may        99         2     -1         0  unknown  no  
2229  unknown   12   may       166         2     -1         0  unknown  no  
3018  unknown   14   may       302         2     -1         0  unknown  no  
5799  unknown   26   may       592         4     -1         0  unknown  no  
6053  unknown   26   may       102        23     -1         0  unknown  no

Se puede ver que hay edades de 123 y mayores que no tienen sentido, se reemplazara por la mediana de las edades restantes.

In [ ]:
# Mediana de los menores o igual a 70
mediana_edad = datos.loc[datos['age'] <= 70, 'age'].median()
mediana_edad

np.float64(40.0)

In [ ]:
datos.loc[datos['age'] > 70, 'age'] = mediana_edad

Ahora si la edad estan en un intervalo logico

In [ ]:
datos["age"].describe()

count    20738.000000
mean        41.001881
std          9.552779
min         20.000000
25%         33.000000
50%         40.000000
75%         49.000000
max         61.000000
Name: age, dtype: float64

## Job

In [ ]:
datos["job"].value_counts()

job
blue-collar       5413
management        3739
technician        3319
admin.            2349
services          2183
retired            841
entrepreneur       752
housemaid          685
self-employed      664
unemployed         474
unknown            159
student            151
administrative       3
Management           2
MANAGEMENT           1
Self-employed        1
Services             1
Retired              1
Name: count, dtype: int64

Se borrara espacios y convirtiendo todo a minuscula

In [ ]:
datos["job"] = datos["job"].astype("object")
datos["job"] = datos["job"].str.strip()
datos["job"] = datos["job"].str.lower()
datos["job"].value_counts()

job
blue-collar       5413
management        3742
technician        3319
admin.            2349
services          2184
retired            842
entrepreneur       752
housemaid          685
self-employed      665
unemployed         474
unknown            159
student            151
administrative       3
Name: count, dtype: int64

In [ ]:
datos[datos["job"]=="administrative"]

age             job   marital  education default  balance housing loan  \
174   60  administrative  divorced  secondary      no       92     yes   no   
175   53  administrative  divorced  secondary      no     1151     yes   no   
376   38  administrative    single  secondary      no      393      no   no   

     contact  day month  duration  campaign  pdays  previous poutcome   y  
174  unknown    5   may       130         1     -1         0  unknown  no  
175  unknown    5   may       412         1     -1         0  unknown  no  
376  unknown    6   may       184         2     -1         0  unknown  no

In [ ]:
datos[datos["job"]=="admin."]

age     job  marital  education default  balance housing loan  \
10      41  admin.     div.  secondary      no      270     yes   no   
11      29  admin.   single  secondary      no      390     yes   no   
16      45  admin.   single    unknown      no       13     yes   no   
25      44  admin.  married  secondary      no     -372     yes   no   
32      60  admin.  married  secondary      no       39     yes  yes   
...    ...     ...      ...        ...     ...      ...     ...  ...   
20675   59  admin.  married   tertiary      no      349      no   no   
20686   56  admin.  married    unknown      no     1167      no   no   
20728   45  admin.  married  secondary     yes     -571      no   no   
20735   51  admin.  married  secondary      no       21      no  yes   
20737   51  admin.  married  secondary      no       79      no   no   

        contact  day month  duration  campaign  pdays  previous poutcome   y  
10      unknown    5   may       222         1     -1         0  unknown  no  
11      unknown    5   may       137         1     -1         0      UNK  no  
16      unknown    5   may        98         1     -1         0  unknown  no  
25      unknown    5   may       172         1     -1         0  unknown  no  
32      unknown    5   may       208         1     -1         0  unknown  no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ..  
20675  cellular   12   aug       124         4     -1         0  unknown  no  
20686  cellular   12   aug       465         9     -1         0  unknown  no  
20728  cellular   13   aug       332         2     -1         0  unknown  no  
20735  cellular   13   aug       157         2     -1         0  unknown  no  
20737  cellular   13   aug        89         2     -1         0  unknown  no  

[2349 rows x 17 columns]

Reemplazare admin. por administrative

In [ ]:
datos["job"] = datos["job"].str.replace("admin.","administrative")

In [ ]:
datos["job"].value_counts()

job
blue-collar       5413
management        3742
technician        3319
administrative    2352
services          2184
retired            842
entrepreneur       752
housemaid          685
self-employed      665
unemployed         474
unknown            159
student            151
Name: count, dtype: int64

## Marital

In [ ]:
datos["marital"].value_counts()

marital
married     13213
single       5002
divorced     2509
div.            7
Single          4
DIVORCED        3
Name: count, dtype: int64

Se eliminara espacios, se convertira a minuscula y se reemplazara algunas categorias para unificar como div. por divorced

In [ ]:
datos["marital"] = datos["marital"].str.strip()
datos["marital"] = datos["marital"].str.lower()

In [ ]:
datos["marital"] = datos["marital"].replace("div.","divorced")

In [ ]:
datos["marital"].value_counts()

marital
married     13213
single       5006
divorced     2519
Name: count, dtype: int64

## Education

In [ ]:
datos["education"].value_counts()

education
secondary    11062
tertiary      4985
primary       3741
unknown        939
SECONDARY        3
Primary          2
UNK              2
sec.             2
Secondary        1
Tertiary         1
Name: count, dtype: int64

Se eliminara espacios, se convertira a minuscula y se reemplazara algunas categorias para unificar como sec. por secondary

In [ ]:
datos["education"] = datos["education"].str.strip()
datos["education"] = datos["education"].str.lower()

In [ ]:
datos["education"].value_counts()

education
secondary    11066
tertiary      4986
primary       3743
unknown        939
sec.             2
unk              2
Name: count, dtype: int64

In [ ]:
datos["education"] = datos["education"].replace("unk","unknown")
datos["education"]  = datos["education"].replace("sec.","secondary")

In [ ]:
datos["education"].value_counts()

education
secondary    11068
tertiary      4986
primary       3743
unknown        941
Name: count, dtype: int64

## Default

In [ ]:
datos["default"].value_counts()

default
no     20217
yes      521
Name: count, dtype: int64

Todo bien en default, dos categorias si y no

## Balance

In [ ]:
datos['balance'].describe()

count     20738.000000
mean       1148.694956
std        4501.391764
min       -8019.000000
25%          30.000000
50%         350.000000
75%        1191.000000
max      527532.000000
Name: balance, dtype: float64

Todo bien balance hay valores negativos y positivos que son aceptables para la variable

## Housing

In [ ]:
datos["housing"].value_counts()

housing
yes    12803
no      7935
Name: count, dtype: int64

Todo bien en housing dos categorias si y no

## Loan

In [ ]:
datos["loan"].value_counts()

loan
no     16589
yes     4139
No         5
YES        2
NO         2
Yes        1
Name: count, dtype: int64

Se eliminara espacios, se convertira a minuscula

In [ ]:
datos["loan"] = datos["loan"].str.strip()
datos["loan"] = datos["loan"].str.lower()

In [ ]:
datos["loan"].value_counts()

loan
no     16596
yes     4142
Name: count, dtype: int64

## Contact

In [ ]:
datos["contact"].value_counts()

contact
unknown      12684
cellular      7237
telephone      817
Name: count, dtype: int64

Todo bien en contacto

## Day

In [ ]:
datos['day'].describe()

count    20738.000000
mean        15.976854
std          8.588358
min          1.000000
25%          8.000000
50%         15.000000
75%         23.000000
max         31.000000
Name: day, dtype: float64

Todo bien en day estan en el rango de 1 a 31

## Month

In [ ]:
datos["month"].value_counts()

month
may    7957
jul    6380
jun    4486
aug    1914
nov       1
Name: count, dtype: int64

Todo bien en month

## Duration

In [ ]:
datos['duration'].describe()

count    20738.000000
mean       259.036358
std        261.616977
min      -1389.000000
25%        103.000000
50%        178.000000
75%        317.000000
max       3881.000000
Name: duration, dtype: float64

Hay duraciones en negativos, se lo convertira en positivo

In [ ]:
datos[datos['duration']<0]

age         job  marital  education default  balance housing loan  \
12   53  technician  married  secondary      no        6     yes   no   
87   41  technician  married  secondary      no     1270     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome    y  
12  unknown    5   may      -517         1     -1         0  unknown   no  
87  unknown    5   may     -1389         1     -1         0  unknown  yes

In [ ]:
datos['duration'] = datos['duration'].abs()

In [ ]:
datos["duration"].describe()

count    20738.000000
mean       259.220176
std        261.434836
min          0.000000
25%        103.000000
50%        178.000000
75%        317.000000
max       3881.000000
Name: duration, dtype: float64

## Campaign

In [ ]:
datos["campaign"].describe()

count    20738.000000
mean         3.089642
std          3.660236
min          1.000000
25%          1.000000
50%          2.000000
75%          3.000000
max         63.000000
Name: campaign, dtype: float64

Todo bien en campaign

## Pdays

In [ ]:
datos["pdays"].value_counts()

pdays
-1      20737
 187        1
Name: count, dtype: int64

Raro que solo haya uno 187 y los otros -1, pero se dejara asi

## Previous

In [ ]:
datos["previous"].value_counts()

previous
0    20737
5        1
Name: count, dtype: int64

Raro que solo haya uno de 5 y los otros 0, pero se dejara asi

## Poutcome

In [ ]:
datos["poutcome"].value_counts()

poutcome
unknown    20733
UNK            4
failure        1
Name: count, dtype: int64

Se reemplazara UNK por unknown

In [ ]:
datos["poutcome"] = datos["poutcome"].str.replace("UNK","unknown")

In [ ]:
datos["poutcome"].value_counts()

poutcome
unknown    20737
failure        1
Name: count, dtype: int64

## y

In [ ]:
datos["y"].value_counts()

y
no     19748
yes      990
Name: count, dtype: int64

Todo bien en y